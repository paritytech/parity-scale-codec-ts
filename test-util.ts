/// <reference lib="deno.unstable"/>

import { dirname, join } from "std/path/win32.ts";
import { assertEquals } from "std/testing/asserts.ts";
import { assertSnapshot } from "std/testing/snapshot.ts";
import { Codec } from "./common.ts";

const file = (path: string) => () => fetch(join(dirname(import.meta.url), path)).then((x) => x.text());
export const files = {
  lipsum: file("lipsum.txt"),
  words: file("words.txt"),
  cargoLock: file("Cargo.lock"),
};

export function testCodec<T>(name: string, codec: Codec<T>, values: T[] | Record<string, T>) {
  for (const key in values) {
    const value = values[key as never] as T;
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key;
    Deno.test(`${name} ${label}`, async (t) => {
      const encoded = codec.encode(value);
      const decoded = codec.decode(encoded);
      assertEquals(decoded, value);
      await assertSnapshot(t, encoded, { serializer: serializeU8A });
    });
  }
}

function serializeU8A(array: Uint8Array) {
  return [...array].map((x) => x.toString(16).padStart(2, "0")).join("\n");
}

export function benchCodec<T>(name: string, codec: Codec<T>, value: T) {
  const encoded = codec.encode(value);
  Deno.bench(`- ${name} (${encoded.length}B) [encode] `, () => {
    codec.encode(value);
  });
  Deno.bench(`  ${name} (${encoded.length}B) [decode]`, () => {
    codec.decode(encoded);
  });
}
