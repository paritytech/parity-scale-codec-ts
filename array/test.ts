import * as $ from "../mod.ts";
import { testCodec } from "../test-util.ts";

testCodec("u8[]", $.array($.u8), [
  [1, 2, 3, 4, 5],
  [6, 7, 8, 9, 10],
  [11, 12, 13, 14, 15],
  [16, 17, 18, 19, 20, 21],
]);

testCodec("[u8; 0]", $.sizedArray(0, $.u8), [[]]);
testCodec("[u8; 1]", $.sizedArray(1, $.u8), [[1]]);
testCodec("[u8; 2]", $.sizedArray(2, $.u8), [[1, 1]]);
testCodec("[u8; 100]", $.sizedArray(100, $.u8), [Array(100).fill(1) as any]);
