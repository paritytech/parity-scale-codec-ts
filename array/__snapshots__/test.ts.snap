export const snapshot = {};

snapshot[`u8[] [ 1, 2, 3, 4, 5 ] 1`] = `
14
01
02
03
04
05
`;

snapshot[`u8[] [ 6, 7, 8, 9, 10 ] 1`] = `
14
06
07
08
09
0a
`;

snapshot[`u8[] [ 11, 12, 13, 14, 15 ] 1`] = `
14
0b
0c
0d
0e
0f
`;

snapshot[`u8[] [ 16, 17, 18, 19, 20, 21 ] 1`] = `
18
10
11
12
13
14
15
`;

snapshot[`[u8; 0] [] 1`] = ``;

snapshot[`[u8; 1] [ 1 ] 1`] = `01`;

snapshot[`[u8; 2] [ 1, 1 ] 1`] = `
01
01
`;

snapshot[`[u8; 100] [
  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
  1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
] 1`] = `
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
01
`;
